File: js/src/uv-harmony.ts
import {SliceCache, uvOfArgb, whiteUvPoint} from './afterimage.js';
import {argbFromXyz, Hct, lstarFromY, sanitizeDegreesDouble, type ColorModel, yFromLstar} from './hct.js';

/** How harmony colors choose rendered tone after u'v' direction is chosen. */
export type HarmonyTonePolicy = 'preserveTone' | 'reflectTone' | 'fitUvChroma' | 'fitHctChroma';

/** Direction of color's chromaticity from the white point, in degrees. */
export function uvHue(color: Hct): number {
  const p = uvOfArgb(color.toInt());
  if (p === null) return 0;
  const [wu, wv] = whiteUvPoint;
  return sanitizeDegreesDouble(Math.atan2(p[1] - wv, p[0] - wu) * 180 / Math.PI);
}

/** Distance of color's chromaticity from the white point in u'v'. */
export function uvChroma(color: Hct): number {
  const p = uvOfArgb(color.toInt());
  if (p === null) return 0;
  const [wu, wv] = whiteUvPoint;
  const du = p[0] - wu, dv = p[1] - wv;
  return Math.sqrt(du * du + dv * dv);
}

/** Color at tone whose chromaticity is uvChroma from white at uvHue. */
export function hctFromUv(
  tone: number,
  hue: number,
  chroma: number,
  model?: ColorModel,
): Hct {
  const [wu, wv] = whiteUvPoint;
  const radians = hue * Math.PI / 180;
  const slice = SliceCache.of(tone);
  const result = slice.nearest(
      wu + chroma * Math.cos(radians),
      wv + chroma * Math.sin(radians));
  const retoned = Hct.from(result.hue, result.chroma, Math.min(100, Math.max(0, tone)));
  return Hct.fromInt(retoned.toInt(), model);
}

const xyzToSrgb = [
  [3.2413774792388685, -1.5376652402851851, -0.49885366846268053],
  [-0.9691452513005321, 1.8758853451067872, 0.04156585616912061],
  [0.05562093689691305, -0.20395524564742123, 1.0571799111220335],
] as const;

function rotateOffset(offset: [number, number], degrees: number): [number, number] {
  const radians = degrees * Math.PI / 180;
  const cos = Math.cos(radians), sin = Math.sin(radians);
  return [offset[0] * cos - offset[1] * sin, offset[0] * sin + offset[1] * cos];
}

function xyzFromUvTone(u: number, v: number, tone: number): [number, number, number] | null {
  if (Math.abs(v) < 1e-9) return null;
  const y = yFromLstar(Math.min(100, Math.max(0, tone)));
  const x = 9 * u * y / (4 * v);
  const z = y * (12 - 3 * u - 20 * v) / (4 * v);
  return [x, y, z];
}

function linearRgbForUvTone(u: number, v: number, tone: number): [number, number, number] | null {
  const xyz = xyzFromUvTone(u, v, tone);
  if (xyz === null) return null;
  const [x, y, z] = xyz;
  return [
    xyzToSrgb[0][0] * x + xyzToSrgb[0][1] * y + xyzToSrgb[0][2] * z,
    xyzToSrgb[1][0] * x + xyzToSrgb[1][1] * y + xyzToSrgb[1][2] * z,
    xyzToSrgb[2][0] * x + xyzToSrgb[2][1] * y + xyzToSrgb[2][2] * z,
  ];
}

function uvToneInSrgb(u: number, v: number, tone: number): boolean {
  const rgb = linearRgbForUvTone(u, v, tone);
  if (rgb === null) return false;
  const epsilon = 1e-7;
  return rgb[0] >= -epsilon && rgb[1] >= -epsilon && rgb[2] >= -epsilon &&
      rgb[0] <= 100 + epsilon && rgb[1] <= 100 + epsilon && rgb[2] <= 100 + epsilon;
}

function fitToneForUv(u: number, v: number, preferredTone: number): number | null {
  if (Math.abs(v) < 1e-9) return null;
  const xPerY = 9 * u / (4 * v);
  const zPerY = (12 - 3 * u - 20 * v) / (4 * v);
  const rPerY = xyzToSrgb[0][0] * xPerY + xyzToSrgb[0][1] + xyzToSrgb[0][2] * zPerY;
  const gPerY = xyzToSrgb[1][0] * xPerY + xyzToSrgb[1][1] + xyzToSrgb[1][2] * zPerY;
  const bPerY = xyzToSrgb[2][0] * xPerY + xyzToSrgb[2][1] + xyzToSrgb[2][2] * zPerY;
  const epsilon = 1e-9;
  if (rPerY < -epsilon || gPerY < -epsilon || bPerY < -epsilon) return null;
  let maxY = 100;
  for (const c of [rPerY, gPerY, bPerY]) {
    if (c > epsilon) maxY = Math.min(maxY, 100 / c);
  }
  const maxTone = lstarFromY(Math.min(100, Math.max(0, maxY)));
  return Math.min(Math.min(100, Math.max(0, preferredTone)), maxTone);
}

function exactUvAtTone(u: number, v: number, tone: number, model?: ColorModel): Hct {
  const xyz = xyzFromUvTone(u, v, tone);
  if (xyz === null) return Hct.from(0, 0, tone, model);
  return Hct.fromInt(argbFromXyz(xyz[0], xyz[1], xyz[2]), model);
}

function targetTone(input: Hct, degrees: number, tonePolicy: HarmonyTonePolicy, u: number, v: number): number {
  if (Math.abs(degrees) < 1e-9 || Math.abs(degrees % 360) < 1e-9) return input.tone;
  switch (tonePolicy) {
    case 'preserveTone': return input.tone;
    case 'reflectTone': return 100 - input.tone;
    case 'fitUvChroma': return fitToneForUv(u, v, input.tone) ?? input.tone;
    case 'fitHctChroma': return input.tone;
  }
}

function rayColorAtScale(tone: number, du: number, dv: number, scale: number, model?: ColorModel): Hct {
  const [wu, wv] = whiteUvPoint;
  const result = SliceCache.of(tone).nearest(wu + scale * du, wv + scale * dv);
  const retoned = Hct.from(result.hue, result.chroma, tone);
  return Hct.fromInt(retoned.toInt(), model);
}

function hctHueFromUvTarget(tone: number, du: number, dv: number, model?: ColorModel): number {
  const exit = SliceCache.of(tone).rayExit(du, dv);
  if (exit === null) return 0;
  return rayColorAtScale(tone, du, dv, Math.min(1, exit[0] * 0.995), model).hue;
}

function isBetterHctChromaFit(candidate: Hct, best: Hct | null, input: Hct): boolean {
  if (best === null) return true;
  const chromaDiff = Math.abs(candidate.chroma - input.chroma);
  const bestChromaDiff = Math.abs(best.chroma - input.chroma);
  if (chromaDiff < bestChromaDiff - 1e-9) return true;
  if (Math.abs(chromaDiff - bestChromaDiff) > 1e-9) return false;
  return Math.abs(candidate.tone - input.tone) < Math.abs(best.tone - input.tone);
}

function fitHctChromaFromUvHue(input: Hct, du: number, dv: number): Hct {
  const targetHue = hctHueFromUvTarget(input.tone, du, dv, input.colorModel);
  let best: Hct | null = null;
  const visit = (tone: number): void => {
    const candidate = Hct.from(targetHue, input.chroma, Math.min(100, Math.max(0, tone)), input.colorModel);
    if (isBetterHctChromaFit(candidate, best, input)) best = candidate;
  };

  for (let tone = 0; tone <= 100; tone += 0.5) visit(tone);
  const coarse = best!.tone;
  for (let tone = coarse - 0.5; tone <= coarse + 0.5; tone += 0.05) visit(tone);
  return best!;
}

function rotated(
  input: Hct,
  offset: [number, number],
  degrees: number,
  sharedStrength: number,
  tonePolicy: HarmonyTonePolicy,
): Hct {
  const [wu, wv] = whiteUvPoint;
  const [rdu, rdv] = rotateOffset(offset, degrees);
  const du = rdu * sharedStrength, dv = rdv * sharedStrength;
  const targetU = wu + du, targetV = wv + dv;
  const tone = targetTone(input, degrees, tonePolicy, targetU, targetV);

  if (tonePolicy === 'fitUvChroma' && uvToneInSrgb(targetU, targetV, tone)) {
    return exactUvAtTone(targetU, targetV, tone, input.colorModel);
  }

  if (tonePolicy === 'fitHctChroma') {
    return fitHctChromaFromUvHue(input, du, dv);
  }

  const slice = SliceCache.of(tone);
  const exit = slice.rayExit(du, dv);
  if (exit === null) return Hct.from(0, 0, tone);
  const s = Math.min(1, exit[0] * 0.995);
  const result = slice.nearest(wu + s * du, wv + s * dv);
  const retoned = Hct.from(result.hue, result.chroma, tone);
  return Hct.fromInt(retoned.toInt(), input.colorModel);
}

/**
 * n evenly-spaced chromatic directions including x. n=2 is complement,
 * n=3 triad, n=4 quad. The set can mix to neutral gray.
 */
export function harmony(
  x: Hct,
  n: number,
  options: {balanced?: boolean, tonePolicy?: HarmonyTonePolicy} = {},
): Hct[] {
  if (n < 2) throw new Error('harmony needs at least 2 colors');
  const tonePolicy = options.tonePolicy ?? 'preserveTone';
  const p = uvOfArgb(x.toInt());
  if (p === null || x.chroma < 1e-4) return Array.from({length: n}, () => x);
  const [wu, wv] = whiteUvPoint;
  const offset: [number, number] = [p[0] - wu, p[1] - wv];
  const angles = Array.from({length: n}, (_, i) => i * 360 / n);
  let shared = 1;
  if (options.balanced === true) {
    for (const a of angles) {
      const [du, dv] = rotateOffset(offset, a);
      const tone = targetTone(x, a, tonePolicy, wu + du, wv + dv);
      const exit = SliceCache.of(tone).rayExit(du, dv);
      if (exit !== null) shared = Math.min(shared, exit[0] * 0.995);
    }
  }
  return angles.map((a) => (a === 0 && options.balanced !== true) ? x : rotated(x, offset, a, shared, tonePolicy));
}

/** Small neighboring steps around x in u'v' direction. */
export function analogous(
  x: Hct,
  options: {count?: number, step?: number, tonePolicy?: HarmonyTonePolicy} = {},
): Hct[] {
  const count = options.count ?? 5;
  const step = options.step ?? 30;
  const tonePolicy = options.tonePolicy ?? 'preserveTone';
  if (count < 1) throw new Error('analogous needs at least 1 color');
  const p = uvOfArgb(x.toInt());
  if (p === null || x.chroma < 1e-4) return Array.from({length: count}, () => x);
  const [wu, wv] = whiteUvPoint;
  const offset: [number, number] = [p[0] - wu, p[1] - wv];
  return Array.from({length: count}, (_, i) => {
    if (i * 2 === count - 1) return x;
    return rotated(x, offset, (i - (count - 1) / 2) * step, 1, tonePolicy);
  });
}

File: lib/effects/uv_harmony.dart
// Modified and maintained by open-source contributors, on behalf of libmonet.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// Color harmony as vector arithmetic around the white point.
///
/// A color is gray plus a chromatic push: in CIE 1976 u'v' (where light
/// mixing is linear) the push is a vector from the white point with a
/// direction ([HctUvCoordinates.uvHue]) and a magnitude
/// ([HctUvCoordinates.uvChroma]). Harmony operations are arithmetic on the
/// direction:
///
///  - complement: +180 degrees ([afterimageComplement])
///  - [harmony]: n evenly spaced directions; the set can mix to gray
///  - [analogous]: small steps to either side
///
/// Direction is the definition; magnitude yields to the gamut. Note that
/// HCT chroma is not comparable across hues (equal chromatic strength can
/// cost C23 in blue and C30 in teal); uvChroma is the comparable measure.
library;

import 'dart:math' as math;

import 'package:libmonet/colorspaces/color_model.dart';
import 'package:libmonet/colorspaces/hct.dart';
import 'package:libmonet/effects/afterimage.dart';

/// u'v' polar coordinates of a color's chromatic push from the white point.
extension HctUvCoordinates on Hct {
  /// Direction of this color's chromaticity from the white point, in
  /// degrees [0, 360). 0 for achromatic colors.
  double get uvHue {
    final p = uvOfArgb(toInt());
    if (p == null) return 0.0;
    final (wu, wv) = whiteUvPoint;
    final degrees =
        math.atan2(p.$2 - wv, p.$1 - wu) * 180.0 / math.pi;
    return (degrees % 360.0 + 360.0) % 360.0;
  }

  /// Distance of this color's chromaticity from the white point in u'v'
  /// (0 = achromatic; sRGB colors reach ~0.28 at vivid red).
  double get uvChroma {
    final p = uvOfArgb(toInt());
    if (p == null) return 0.0;
    final (wu, wv) = whiteUvPoint;
    final du = p.$1 - wu, dv = p.$2 - wv;
    return math.sqrt(du * du + dv * dv);
  }
}

/// Color at [tone] whose chromaticity sits [uvChroma] from the white point
/// in direction [uvHue] (degrees), clamped to the sRGB gamut.
Hct hctFromUv(
  double tone,
  double uvHue,
  double uvChroma, {
  ColorModel model = ColorModel.kDefault,
}) {
  final (wu, wv) = whiteUvPoint;
  final radians = uvHue * math.pi / 180.0;
  final slice = SliceCache.of(tone);
  final result = slice.nearest(
    wu + uvChroma * math.cos(radians),
    wv + uvChroma * math.sin(radians),
  );
  final retoned = Hct.from(result.hue, result.chroma, tone.clamp(0.0, 100.0));
  return Hct.fromInt(retoned.toInt(), model: model);
}

Hct _rotated(
  Hct input,
  SliceCache slice,
  (double, double) offset,
  double degrees,
  double sharedStrength,
) {
  final (wu, wv) = whiteUvPoint;
  final radians = degrees * math.pi / 180.0;
  final cos = math.cos(radians), sin = math.sin(radians);
  final du = offset.$1 * cos - offset.$2 * sin;
  final dv = offset.$1 * sin + offset.$2 * cos;
  final exit = slice.rayExit(du, dv);
  if (exit == null) return Hct.from(0.0, 0.0, input.tone);
  final s = math.min(sharedStrength, exit.$1 * 0.995);
  final result = slice.nearest(wu + s * du, wv + s * dv);
  final retoned = Hct.from(result.hue, result.chroma, input.tone);
  return Hct.fromInt(retoned.toInt(), model: input.colorModel);
}

/// [n] colors at [x]'s tone whose chromatic directions are evenly spaced
/// around the white point, starting at [x]. Because the directions cancel,
/// an additive mixture of the set (at some ratio) is neutral gray: this is
/// the checkable property behind complement (n=2), triad (n=3), and
/// quad (n=4). x itself is the first element.
///
/// By default each color is as vivid as the gamut allows in its direction.
/// With [balanced], all are capped at the weakest direction's strength for
/// an even ensemble (equal [HctUvCoordinates.uvChroma], not equal HCT
/// chroma; the two differ by hue).
///
/// Achromatic inputs return n copies of themselves: no direction to spread.
List<Hct> harmony(Hct x, int n, {bool balanced = false}) {
  assert(n >= 2, 'harmony needs at least 2 colors');
  final p = uvOfArgb(x.toInt());
  if (p == null || x.chroma < 1e-4) return List.filled(n, x);
  final (wu, wv) = whiteUvPoint;
  final offset = (p.$1 - wu, p.$2 - wv);
  final slice = SliceCache.of(x.tone);
  final angles = [for (var i = 0; i < n; i++) i * 360.0 / n];
  var shared = 1.0;
  if (balanced) {
    for (final a in angles) {
      final radians = a * math.pi / 180.0;
      final du = offset.$1 * math.cos(radians) - offset.$2 * math.sin(radians);
      final dv = offset.$1 * math.sin(radians) + offset.$2 * math.cos(radians);
      final exit = slice.rayExit(du, dv);
      if (exit != null) shared = math.min(shared, exit.$1 * 0.995);
    }
  }
  return [
    for (final a in angles)
      if (a == 0.0 && !balanced)
        x // the seed itself, untouched
      else
        _rotated(x, slice, offset, a, shared),
  ];
}

/// [count] colors at [x]'s tone: [x] flanked by neighbors [step] degrees
/// apart in chromatic direction (u'v' angle around the white point), at
/// [x]'s chromatic strength. For odd [count], x is the middle element.
///
/// Unlike [harmony] sets, analogous colors all lean the same way from gray
/// and cannot mix to neutral; this is a nearness operation, not a balance
/// operation. [step] is a design knob, not physics: 30 degrees of u'v'
/// direction per step is the default.
List<Hct> analogous(Hct x, {int count = 5, double step = 30.0}) {
  assert(count >= 1, 'analogous needs at least 1 color');
  final p = uvOfArgb(x.toInt());
  if (p == null || x.chroma < 1e-4) return List.filled(count, x);
  final (wu, wv) = whiteUvPoint;
  final offset = (p.$1 - wu, p.$2 - wv);
  final slice = SliceCache.of(x.tone);
  return [
    for (var i = 0; i < count; i++)
      if (i * 2 == count - 1)
        x // the seed itself, untouched
      else
        _rotated(x, slice, offset, (i - (count - 1) / 2.0) * step, 1.0),
  ];
}
